File: src/test/java/com/hedera/hashgraph/sdk/HookStoreTransactionTest.java
// SPDX-License-Identifier: Apache-2.0
package com.hedera.hashgraph.sdk;

import static org.assertj.core.api.Assertions.assertThat;
import static org.junit.jupiter.api.Assertions.assertThrows;

import com.hedera.hashgraph.sdk.proto.HookStoreTransactionBody;
import com.hedera.hashgraph.sdk.proto.TransactionBody;
import java.time.Instant;
import java.util.Arrays;
import java.util.List;
import org.junit.jupiter.api.Test;

public class HookStoreTransactionTest {

    private static final PrivateKey TEST_PRIVATE_KEY = PrivateKey.fromString(
            "302e020100300506032b657004220420db484b828e64b2d8f12ce3c0a0e93a0b8cce7af1bb8f39c97732394482538e10");

    private static final HookId TEST_HOOK_ID = new HookId(new HookEntityId(AccountId.fromString("0.0.5006")), 42);

    private static final List<EvmHookStorageUpdate> TEST_UPDATES = List.of(
            new EvmHookStorageUpdate.EvmHookStorageSlot(new byte[] {0x01}, new byte[] {0x02}),
            new EvmHookStorageUpdate.EvmHookStorageSlot(new byte[] {0x03}, new byte[] {0x04}));

    final Instant TEST_VALID_START = Instant.ofEpochSecond(1554158542);

    private HookStoreTransaction spawnTestTransaction() {
        return new HookStoreTransaction()
                .setNodeAccountIds(Arrays.asList(AccountId.fromString("0.0.5005"), AccountId.fromString("0.0.5006")))
                .setTransactionId(TransactionId.withValidStart(AccountId.fromString("0.0.5006"), TEST_VALID_START))
                .setHookId(TEST_HOOK_ID)
                .setStorageUpdates(TEST_UPDATES)
                .setMaxTransactionFee(new Hbar(1))
                .freeze()
                .sign(TEST_PRIVATE_KEY);
    }

    @Test
    void bytesRoundTripNoSetters() throws Exception {
        var tx = new HookStoreTransaction();
        var tx2 = Transaction.fromBytes(tx.toBytes());
        assertThat(tx2.toString()).isEqualTo(tx.toString());
    }

    @Test
    void bytesRoundTripWithSetters() throws Exception {
        var tx = spawnTestTransaction();
        var tx2 = Transaction.fromBytes(tx.toBytes());
        assertThat(tx2.toString()).isEqualTo(tx.toString());
        assertThat(tx2).isInstanceOf(HookStoreTransaction.class);
    }

    // HookStoreTransaction is not schedulable; no scheduled mapping exists.

    @Test
    void constructFromTransactionBodyProtobuf() {
        var hookBody = HookStoreTransactionBody.newBuilder()
                .setHookId(TEST_HOOK_ID.toProtobuf())
                .addAllStorageUpdates(TEST_UPDATES.stream()
                        .map(EvmHookStorageUpdate::toProtobuf)
                        .toList())
                .build();

        var txBody = TransactionBody.newBuilder().setHookStore(hookBody).build();
        var tx = new HookStoreTransaction(txBody);

        assertThat(tx.getHookId()).isEqualTo(TEST_HOOK_ID);
        assertThat(tx.getStorageUpdates()).hasSize(TEST_UPDATES.size());
    }

    @Test
    void settersAndFrozenBehavior() {
        var tx = new HookStoreTransaction().setHookId(TEST_HOOK_ID).setStorageUpdates(TEST_UPDATES);

        assertThat(tx.getHookId()).isEqualTo(TEST_HOOK_ID);
        assertThat(tx.getStorageUpdates()).isEqualTo(TEST_UPDATES);

        var frozen = spawnTestTransaction();
        assertThrows(IllegalStateException.class, () -> frozen.setHookId(TEST_HOOK_ID));
        assertThrows(IllegalStateException.class, () -> frozen.setStorageUpdates(TEST_UPDATES));
        assertThrows(IllegalStateException.class, () -> frozen.addStorageUpdate(TEST_UPDATES.get(0)));
    }

    @Test
    void validateChecksumsWithNullHookIdDoesNotThrow() throws Exception {
        var client = Client.forTestnet();
        var tx = new HookStoreTransaction();

        // Should not throw when hookId is null
        tx.validateChecksums(client);
    }

    @Test
    void validateChecksumsWithAccountHookIdValidatesAccountId() throws Exception {
        var client = Client.forTestnet();
        var accountId = AccountId.fromString("0.0.1234");
        var hookId = new HookId(new HookEntityId(accountId), 1L);
        var tx = new HookStoreTransaction().setHookId(hookId);

        // Should not throw with valid account ID
        tx.validateChecksums(client);
    }

    @Test
    void validateChecksumsWithContractHookIdValidatesContractId() throws Exception {
        var client = Client.forTestnet();
        var contractId = ContractId.fromString("0.0.5678");
        var hookId = new HookId(new HookEntityId(contractId), 2L);
        var tx = new HookStoreTransaction().setHookId(hookId);

        // Should not throw with valid contract ID
        tx.validateChecksums(client);
    }

    @Test
    void validateChecksumsWithInvalidAccountIdThrows() throws Exception {
        var client = Client.forTestnet();
        // Create an account ID with invalid checksum (using a known bad checksum from AccountIdTest)
        var accountId = AccountId.fromString("0.0.123-ntjli");
        var hookId = new HookId(new HookEntityId(accountId), 3L);
        var tx = new HookStoreTransaction().setHookId(hookId);

        // Should throw BadEntityIdException for invalid checksum
        assertThrows(BadEntityIdException.class, () -> tx.validateChecksums(client));
    }

    @Test
    void validateChecksumsWithInvalidContractIdThrows() throws Exception {
        var client = Client.forTestnet();
        // Create a contract ID with invalid checksum (using a known bad checksum)
        var contractId = ContractId.fromString("0.0.123-ntjli");
        var hookId = new HookId(new HookEntityId(contractId), 4L);
        var tx = new HookStoreTransaction().setHookId(hookId);

        // Should throw BadEntityIdException for invalid checksum
        assertThrows(BadEntityIdException.class, () -> tx.validateChecksums(client));
    }
}

File: src/Hedera.Hashgraph.Tests/HookStoreTransactionTest.cs
// SPDX-License-Identifier: Apache-2.0
using Org.Assertj.Core.Api.Assertions;
using Org.Junit.Jupiter.Api.Assertions;
using Com.Hedera.Hashgraph.Sdk.Proto;
using Java.Time;
using Java.Util;
using Org.Junit.Jupiter.Api;
using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.Linq;
using System.Text;

namespace Com.Hedera.Hashgraph.Sdk
{
    public class HookStoreTransactionTest
    {
        private static readonly PrivateKey TEST_PRIVATE_KEY = PrivateKey.FromString("302e020100300506032b657004220420db484b828e64b2d8f12ce3c0a0e93a0b8cce7af1bb8f39c97732394482538e10");
        private static readonly HookId TEST_HOOK_ID = new HookId(new HookEntityId(AccountId.FromString("0.0.5006")), 42);
        private static readonly IList<EvmHookStorageUpdate> TEST_UPDATES = List.Of(new EvmHookStorageSlot(new byte[] { 0x01 }, new byte[] { 0x02 }), new EvmHookStorageSlot(new byte[] { 0x03 }, new byte[] { 0x04 }));
        readonly Instant TEST_VALID_START = Instant.OfEpochSecond(1554158542);
        private HookStoreTransaction SpawnTestTransaction()
        {
            return new HookStoreTransaction().SetNodeAccountIds(Arrays.AsList(AccountId.FromString("0.0.5005"), AccountId.FromString("0.0.5006"))).SetTransactionId(TransactionId.WithValidStart(AccountId.FromString("0.0.5006"), TEST_VALID_START)).SetHookId(TEST_HOOK_ID).SetStorageUpdates(TEST_UPDATES).SetMaxTransactionFee(new Hbar(1)).Freeze().Sign(TEST_PRIVATE_KEY);
        }

        virtual void BytesRoundTripNoSetters()
        {
            var tx = new HookStoreTransaction();
            var tx2 = Transaction.FromBytes(tx.ToBytes());
            Assert.Equal(tx2.ToString(), tx.ToString());
        }

        virtual void BytesRoundTripWithSetters()
        {
            var tx = SpawnTestTransaction();
            var tx2 = Transaction.FromBytes(tx.ToBytes());
            Assert.Equal(tx2.ToString(), tx.ToString());
            Assert.IsType<HookStoreTransaction>(tx2);
        }

        // HookStoreTransaction is not schedulable; no scheduled mapping exists.
        virtual void ConstructFromTransactionBodyProtobuf()
        {
            var hookBody = HookStoreTransactionBody.NewBuilder().SetHookId(TEST_HOOK_ID.ToProtobuf()).AddAllStorageUpdates(TEST_UPDATES.Stream().Map(EvmHookStorageUpdate.ToProtobuf()).ToList()).Build();
            var txBody = TransactionBody.NewBuilder().SetHookStore(hookBody).Build();
            var tx = new HookStoreTransaction(txBody);
            Assert.Equal(tx.GetHookId(), TEST_HOOK_ID);
            AssertThat(tx.GetStorageUpdates()).HasSize(TEST_UPDATES.Count);
        }

        virtual void SettersAndFrozenBehavior()
        {
            var tx = new HookStoreTransaction().SetHookId(TEST_HOOK_ID).SetStorageUpdates(TEST_UPDATES);
            Assert.Equal(tx.GetHookId(), TEST_HOOK_ID);
            Assert.Equal(tx.GetStorageUpdates(), TEST_UPDATES);
            var frozen = SpawnTestTransaction();
            await Assert.ThrowsAsync<InvalidOperationException>(() => frozen.SetHookId(TEST_HOOK_ID));
            await Assert.ThrowsAsync<InvalidOperationException>(() => frozen.SetStorageUpdates(TEST_UPDATES));
            await Assert.ThrowsAsync<InvalidOperationException>(() => frozen.AddStorageUpdate(TEST_UPDATES[0]));
        }

        virtual void ValidateChecksumsWithNullHookIdDoesNotThrow()
        {
            var client = Client.ForTestnet();
            var tx = new HookStoreTransaction();

            // Should not throw when hookId is null
            tx.ValidateChecksums(client);
        }

        virtual void ValidateChecksumsWithAccountHookIdValidatesAccountId()
        {
            var client = Client.ForTestnet();
            var accountId = AccountId.FromString("0.0.1234");
            var hookId = new HookId(new HookEntityId(accountId), 1);
            var tx = new HookStoreTransaction().SetHookId(hookId);

            // Should not throw with valid account ID
            tx.ValidateChecksums(client);
        }

        virtual void ValidateChecksumsWithContractHookIdValidatesContractId()
        {
            var client = Client.ForTestnet();
            var contractId = ContractId.FromString("0.0.5678");
            var hookId = new HookId(new HookEntityId(contractId), 2);
            var tx = new HookStoreTransaction().SetHookId(hookId);

            // Should not throw with valid contract ID
            tx.ValidateChecksums(client);
        }

        virtual void ValidateChecksumsWithInvalidAccountIdThrows()
        {
            var client = Client.ForTestnet();

            // Create an account ID with invalid checksum (using a known bad checksum from AccountIdTest)
            var accountId = AccountId.FromString("0.0.123-ntjli");
            var hookId = new HookId(new HookEntityId(accountId), 3);
            var tx = new HookStoreTransaction().SetHookId(hookId);

            // Should throw BadEntityIdException for invalid checksum
            await Assert.ThrowsAsync<BadEntityIdException>(() => tx.ValidateChecksums(client));
        }

        virtual void ValidateChecksumsWithInvalidContractIdThrows()
        {
            var client = Client.ForTestnet();

            // Create a contract ID with invalid checksum (using a known bad checksum)
            var contractId = ContractId.FromString("0.0.123-ntjli");
            var hookId = new HookId(new HookEntityId(contractId), 4);
            var tx = new HookStoreTransaction().SetHookId(hookId);

            // Should throw BadEntityIdException for invalid checksum
            await Assert.ThrowsAsync<BadEntityIdException>(() => tx.ValidateChecksums(client));
        }
    }
}